worlds/osrs/__init__.py

```diff
@@ -0,0 +1,616 @@
+        starting_entrance.connect(self.region_name_to_data[starting_area_region])
+
+        # Create entrances between regions
+        for region_row in self.region_rows:
+            region = self.region_name_to_data[region_row.name]
+            for outbound_region_name in region_row.connections:
+                entrance = region.create_exit(f"{region_row.name}->{outbound_region_name.replace('*', '')}")
+                if "*" not in outbound_region_name:
+                    item_name = self.region_rows_by_name[outbound_region_name].itemReq
+                    if "*" not in item_name:
+                        entrance.access_rule = lambda state, item_name=item_name: state.has(item_name, self.player)
+                    else:
+                        self.generate_special_rules_for(entrance, region_row, outbound_region_name)
+                else:
+                    self.generate_special_rules_for(entrance, region_row, outbound_region_name)
+                entrance.connect(self.region_name_to_data[outbound_region_name.replace('*', '')])
+            for resource_region in region_row.resources:
+                if resource_region != "":
+                    entrance = region.create_exit(f"{region_row.name}->{resource_region.replace('*', '')}")
+                    if "*" not in resource_region:
+                        entrance.connect(self.region_name_to_data[resource_region])
+                    else:
+                        self.generate_special_rules_for(entrance, region_row, resource_region)
+                        entrance.connect(self.region_name_to_data[resource_region.replace('*', '')])
+
+        self.roll_locations()
+
+    def generate_special_rules_for(self, entrance, region_row, outbound_region_name):
+        # print(f"Special rules required to access region {outbound_region_name} from {region_row.name}")
+        if outbound_region_name == "Cook's Guild":
+            item_name = self.region_rows_by_name[outbound_region_name].itemReq.replace('*', '')
+            entrance.access_rule = lambda state: state.has(item_name, self.player) and\
+                self.can_reach_skill(state, "cooking", 32)
+            return
+        if outbound_region_name == "Crafting Guild":
+            item_name = self.region_rows_by_name[outbound_region_name].itemReq.replace('*', '')
+            entrance.access_rule = lambda state:  state.has(item_name, self.player) and\
+                self.can_reach_skill(state, "crafting", 40)
+            return
+        if outbound_region_name == "Corsair Cove":
+            item_name = self.region_rows_by_name[outbound_region_name].itemReq.replace('*', '')
+            # Need to be able to start Corsair Curse in addition to having the item
+            entrance.access_rule = lambda state: state.has(item_name, self.player) and \
+                                                 state.can_reach(RegionNames.Falador_Farm, None, self.player)
```

Comment:
**CAREFUL!!!** You are not allowed to perform a **Region**.can_reach from inside an **Entrance**.access_rule, unless you use `register_indirect_condition`

Otherwise, it will lead to generation failures!

Reply:
(Also, why `None` anyway, instead of `"Region"`? :O)